In [1]:
#Importing the libraries
import pandas as pd
import numpy as np
# the url
file = "https://raw.githubusercontent.com/jtopor/DAV-5400/master/Project1/hflights.csv"
# the data frame
df = pd.read_csv(file)

# Model Goal, Dependant and independant variables
Before we start with the feature engineering process we have to first establish our goals
So we need to first identify what the dependent features are and the independent features to create a clear picture of what type of feature engineering is needed and direct our efforts in the right direction.

- So the goal is to determenine whether or not a fligh has been cancelled given a group of features.

so up till now in this context the Dependant and independant variables are as follows:
- dependent feature(s) : Cancelled
- Independant features: all except the Cancelled coloumn

In [2]:
df.shape

(20000, 21)

In [3]:
df[df['Cancelled']==1].head()

Year  Month  DayofMonth  DayOfWeek  DepTime  ArrTime UniqueCarrier  \
194  2011      1          24          1      NaN      NaN            AA   
210  2011      1           9          7      NaN      NaN            AA   
323  2011      1          11          2      NaN      NaN            B6   
335  2011      1          19          3      NaN      NaN            B6   
347  2011      1          27          4      NaN      NaN            B6   

     FlightNum TailNum  ActualElapsedTime  ...  ArrDelay  DepDelay  Origin  \
194       1700  N3BGAA                NaN  ...       NaN       NaN     IAH   
210       1820  N4XCAA                NaN  ...       NaN       NaN     IAH   
323        624  N537JB                NaN  ...       NaN       NaN     HOU   
335        624  N504JB                NaN  ...       NaN       NaN     HOU   
347        624  N569JB                NaN  ...       NaN       NaN     HOU   

    Dest Distance  TaxiIn  TaxiOut  Cancelled  CancellationCode Diverted  
194  MIA      964     NaN      NaN          1                 A        0  
210  DFW      224     NaN      NaN          1                 B        0  
323  JFK     1428     NaN      NaN          1                 B        0  
335  JFK     1428     NaN      NaN          1                 A        0  
347  JFK     1428     NaN      NaN          1                 B        0  

[5 rows x 21 columns]

3904

Let's examine the dependant feature it seems like the values present in the Cancelled column/feature are 1 & 0 from which you can only deduce that it is a binary classification machine learning problem 

In [5]:
df['Cancelled'].unique()

array([0, 1], dtype=int64)

So now that we have established our goals we turn and assigned the Dependant and independant variable leaving us with 1 dependent feature vector and 20 independent feature vectors, we turn our attention to feature engineering and the first feature engineering method we are going to use is 


## 1] Removal of unused attributes/features, Can we ?
Yes, we can here is how and why

  we will follow a black listing approach where we list all the features deemed to be unhelpful towards our statistical model.
I believe the following features are not essential to our goal.


- CancellationCode: reason for cancellation: A = carrier, B = weather, C = national air system, D = security  
- TailNum: airplane tail number 
- FlightNum: : flight number 
- TaxiIn: taxi in time in minutes  
- TaxiOut: taxi out time in minutes  
- Diverted: diverted indicator: 1 = Yes, 0 = N

##### The reason i excecluded Cancellation code is because it not logically sound to try and predict flight cancellation while having Cancellation Code as this a piece of information that we are not suppose to have before hand.

###### TailNum and FlightNum are two forms of ID that are so unique to the extent that it would not help generalize our statistical model, so keeping could actually harm the model and lead to overfitting, However i kept carrier unique id and the reason for that is because the specific airline could give our model an idea if it is going to be cancelled or not and needs to be empircally tested in the model development section.

##### Based on my domain Knowledge , how much time you spend within a taxt doesnot affect whether or not a flight is cancelled, it is more common sense that domain knowledge really.

###### Diverted, if the plane diverted it means that it wasn't cancelled. this is also not logically sound to keep

Note: the reason i chose to do this first is because i would like to exert effort only on the ones that i will be using for predictions as it is always better to define your scope

In [6]:
df.head(1)

Year  Month  DayofMonth  DayOfWeek  DepTime  ArrTime UniqueCarrier  \
0  2011      1           1          6   1400.0   1500.0            AA   

   FlightNum TailNum  ActualElapsedTime  ...  ArrDelay  DepDelay  Origin Dest  \
0        428  N576AA               60.0  ...     -10.0       0.0     IAH  DFW   

  Distance  TaxiIn  TaxiOut  Cancelled  CancellationCode Diverted  
0      224     7.0     13.0          0               NaN        0  

[1 rows x 21 columns]

In [7]:
data=df.iloc[:,:16]
X=data.drop(['FlightNum'], axis=1)
X.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  ArrTime UniqueCarrier TailNum  \
0  2011      1           1          6   1400.0   1500.0            AA  N576AA   
1  2011      1           2          7   1401.0   1501.0            AA  N557AA   
2  2011      1           3          1   1352.0   1502.0            AA  N541AA   
3  2011      1           4          2   1403.0   1513.0            AA  N403AA   
4  2011      1           5          3   1405.0   1507.0            AA  N492AA   

   ActualElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  
0               60.0     40.0     -10.0       0.0    IAH  DFW       224  
1               60.0     45.0      -9.0       1.0    IAH  DFW       224  
2               70.0     48.0      -8.0      -8.0    IAH  DFW       224  
3               70.0     39.0       3.0       3.0    IAH  DFW       224  
4               62.0     44.0      -3.0       5.0    IAH  DFW       224

## 2. Combining Sparse Classifications, Can we after the removal?
we turn our attention to the categorical features left
 
 Unique Carrier
 
 Dest

 Origin 
<br>
we fist check if they have missing values, They appear to not have any missing values, however there are missing values in the data that need to be dealt with later on outside the scope of this step.

In [8]:
X.isnull().sum(axis = 0)

Year                   0
Month                  0
DayofMonth             0
DayOfWeek              0
DepTime              215
ArrTime              234
UniqueCarrier          0
TailNum              108
ActualElapsedTime    262
AirTime              262
ArrDelay             262
DepDelay             215
Origin                 0
Dest                   0
Distance               0
dtype: int64

In [9]:
print(X['UniqueCarrier'].unique())
print('\n')
print('\n')
print(X['Origin'].unique())
print('\n')
print('\n')
print(X['Dest'].unique())

['AA' 'AS' 'B6' 'CO' 'DL' 'OO' 'UA' 'US' 'WN' 'EV' 'F9' 'FL' 'MQ' 'XE']




['IAH' 'HOU']




['DFW' 'MIA' 'SEA' 'JFK' 'HNL' 'MSY' 'SAT' 'AUS' 'LAX' 'DEN' 'EWR' 'ORD'
 'ONT' 'DCA' 'SFO' 'LAS' 'TPA' 'PDX' 'SJU' 'PHX' 'BWI' 'LGA' 'CLE' 'RDU'
 'BOS' 'IND' 'SAN' 'IAD' 'SJC' 'TUL' 'MFE' 'PIT' 'SLC' 'SNA' 'MCO' 'MSP'
 'EGE' 'ATL' 'PHL' 'DTW' 'FLL' 'CLT' 'RSW' 'MCI' 'SMF' 'ELP' 'OKC' 'ABQ'
 'TUS' 'PBI' 'OMA' 'HDN' 'GUC' 'MTJ' 'BHM' 'CMH' 'COS' 'GRR' 'ASE' 'MKE'
 'BNA' 'CRP' 'DAL' 'ECP' 'HRL' 'JAN' 'JAX' 'LIT' 'MAF' 'MDW' 'OAK' 'STL'
 'MEM' 'CVG' 'SDF' 'BTR' 'RIC' 'LRD' 'ORF' 'PNS' 'CRW' 'XNA' 'AEX' 'BRO'
 'DAY' 'GPT' 'GSO' 'SAV' 'ICT' 'HSV' 'AVL' 'LEX' 'LFT' 'LCH' 'LBB' 'DSM'
 'CHS' 'CAE' 'GSP' 'SHV' 'TYS' 'MOB' 'VPS' 'AMA' 'GRK' 'RNO']


After investigating the unique categorical values for Unique and origin coloumn i have decided to keep them as they are as they cannot be grouped, however it is would be good idea to put a pin on destination if we ever encounter a problem in the future with our model and investigate whether or not we could perform a grouping although i highly doubt that that would be the a problem

Conclusion.
No need for the combining sparse elements however we do need to impute and add the missing values by other numerical values (Mean,median...etc), So let us go ahead and perform the third method of feature engineering which is creating dummy variables

## 3] Creating Dummy Variables, Should We?
Yes, we absoultely should and here is how.
- all the categorical variables that are present need to be changed into numerical values that donot allow any form precedence for the machine learning model, that is because we want our model to be able to use it and to be able to use fairly

In [10]:
encoded_columns = pd.get_dummies(X['UniqueCarrier'])
X = X.join(encoded_columns).drop('UniqueCarrier', axis=1)

encoded_columns = pd.get_dummies(X['Origin'])
X = X.join(encoded_columns).drop('Origin', axis=1)

encoded_columns = pd.get_dummies(X['Dest'])
X = X.join(encoded_columns).drop('Dest', axis=1)

In [11]:
X.iloc[:,12:].head()

AA  AS  B6  CO  DL  EV  F9  FL  MQ  OO  ...  SLC  SMF  SNA  STL  TPA  TUL  \
0   1   0   0   0   0   0   0   0   0   0  ...    0    0    0    0    0    0   
1   1   0   0   0   0   0   0   0   0   0  ...    0    0    0    0    0    0   
2   1   0   0   0   0   0   0   0   0   0  ...    0    0    0    0    0    0   
3   1   0   0   0   0   0   0   0   0   0  ...    0    0    0    0    0    0   
4   1   0   0   0   0   0   0   0   0   0  ...    0    0    0    0    0    0   

   TUS  TYS  VPS  XNA  
0    0    0    0    0  
1    0    0    0    0  
2    0    0    0    0  
3    0    0    0    0  
4    0    0    0    0  

[5 rows x 122 columns]

In [15]:
X

Year  Month  DayofMonth  DayOfWeek  DepTime  ArrTime TailNum  \
0      2011      1           1          6   1400.0   1500.0  N576AA   
1      2011      1           2          7   1401.0   1501.0  N557AA   
2      2011      1           3          1   1352.0   1502.0  N541AA   
3      2011      1           4          2   1403.0   1513.0  N403AA   
4      2011      1           5          3   1405.0   1507.0  N492AA   
5      2011      1           6          4   1359.0   1503.0  N262AA   
6      2011      1           7          5   1359.0   1509.0  N493AA   
7      2011      1           8          6   1355.0   1454.0  N477AA   
8      2011      1           9          7   1443.0   1554.0  N476AA   
9      2011      1          10          1   1443.0   1553.0  N504AA   
10     2011      1          11          2   1429.0   1539.0  N565AA   
11     2011      1          12          3   1419.0   1515.0  N577AA   
12     2011      1          13          4   1358.0   1501.0  N476AA   
13     2011      1          14          5   1357.0   1504.0  N552AA   
14     2011      1          15          6   1359.0   1459.0  N462AA   
15     2011      1          16          7   1359.0   1509.0  N555AA   
16     2011      1          17          1   1530.0   1634.0  N518AA   
17     2011      1          18          2   1408.0   1508.0  N507AA   
18     2011      1          19          3   1356.0   1503.0  N523AA   
19     2011      1          20          4   1507.0   1622.0  N425AA   
20     2011      1          21          5   1357.0   1459.0  N251AA   
21     2011      1          22          6   1355.0   1456.0  N551AA   
22     2011      1          23          7   1356.0   1501.0  N479AA   
23     2011      1          24          1   1356.0   1513.0  N531AA   
24     2011      1          25          2   1352.0   1452.0  N561AA   
25     2011      1          26          3   1353.0   1455.0  N541AA   
26     2011      1          27          4   1356.0   1458.0  N512AA   
27     2011      1          28          5   1359.0   1505.0  N4UBAA   
28     2011      1          29          6   1355.0   1455.0  N491AA   
29     2011      1          30          7   1359.0   1456.0  N561AA   
...     ...    ...         ...        ...      ...      ...     ...   
19970  2011      2          25          5    916.0   1004.0  N14645   
19971  2011      2          25          5   1850.0   2148.0  N57863   
19972  2011      2          25          5   1437.0   1614.0  N17233   
19973  2011      2          25          5    728.0   1037.0  N17233   
19974  2011      2          25          5   1900.0   1945.0  N17719   
19975  2011      2          25          5   1201.0   1356.0  N73406   
19976  2011      2          25          5   1328.0   1515.0  N75433   
19977  2011      2          25          5   1612.0   1717.0  N37255   
19978  2011      2          25          5    847.0    946.0  N13750   
19979  2011      2          25          5   1214.0   1353.0  N23707   
19980  2011      2          25          5   2000.0     14.0  N21723   
19981  2011      2          25          5   2051.0   2300.0  N79279   
19982  2011      2          25          5   1100.0   1326.0  N57864   
19983  2011      2          25          5   1602.0   1914.0  N77867   
19984  2011      2          25          5   2035.0   2329.0  N75433   
19985  2011      2          25          5    732.0    842.0  N12221   
19986  2011      2          25          5   1754.0   2154.0  N33294   
19987  2011      2          25          5    914.0   1025.0  N14639   
19988  2011      2          25          5   1616.0   2001.0  N54241   
19989  2011      2          25          5   1734.0   1910.0  N24211   
19990  2011      2          25          5   1029.0   1341.0  N33294   
19991  2011      2          25          5   1451.0   1645.0  N14231   
19992  2011      2          25          5   1903.0   2034.0  N16648   
19993  2011      2          25          5    908.0   1137.0  N32404   
19994  2011      2          25 

# 4 . Creating interaction features
I believe there should be binning when it comes to distance as there are a plethora of bins and it would be a lot more useful to group them between 127 to 3904 , we should perhaps use pd.cut() to establish the binning as we learned in the grouping lectures i believe that would be the best approach to go about it.




for example

In [23]:
from scipy.stats.mstats import mquantiles
mquantiles(X.Distance)

array([ 351.,  787., 1034.])

# 5.Creating Indicators
I think we should create a new column with the season which we will derive from the time-related columns which will tell us in which of the four-season where we, I believe that will have a huge impact on the model as there is the season where the weather is just to be bad.our new created indicator should be a season and it would be classified as 1,2,3,4 for each of the seasons

In [ ]:
from datetime import date, datetime

Y = 2000 # dummy leap year to allow input X-02-29 (leap day)
seasons = [('winter', (date(Y,  1,  1),  date(Y,  3, 20))),
           ('spring', (date(Y,  3, 21),  date(Y,  6, 20))),
           ('summer', (date(Y,  6, 21),  date(Y,  9, 22))),
           ('autumn', (date(Y,  9, 23),  date(Y, 12, 20))),
           ('winter', (date(Y, 12, 21),  date(Y, 12, 31)))]

def get_season(now):
    if isinstance(now, datetime):
        now = now.date()
    now = now.replace(year=Y)
    return next(season for season, (start, end) in seasons
                if start <= now <= end)

print(get_season(date.today()))

refrences:
https://stackoverflow.com/questions/16139306/determine-season-given-timestamp-in-python-using-datetime